diff -up evolution-mapi-0.32.2/src/libexchangemapi/exchange-mapi-cal-utils.c.crash-setting-props evolution-mapi-0.32.2/src/libexchangemapi/exchange-mapi-cal-utils.c
--- evolution-mapi-0.32.2/src/libexchangemapi/exchange-mapi-cal-utils.c.crash-setting-props	2011-02-07 04:45:47.000000000 +0100
+++ evolution-mapi-0.32.2/src/libexchangemapi/exchange-mapi-cal-utils.c	2013-05-29 18:53:07.728612389 +0200
@@ -1342,12 +1342,17 @@ exchange_mapi_cal_utils_write_props_cb (
 	flag32 = prop ? get_imp_prop_from_priority (icalproperty_get_priority (prop)) : IMPORTANCE_NORMAL;
 	set_value (PR_IMPORTANCE, &flag32);
 
-	set_value (PR_SENT_REPRESENTING_NAME_UNICODE, cbdata->ownername);
-	set_value (PR_SENT_REPRESENTING_ADDRTYPE_UNICODE, cbdata->owneridtype);
-	set_value (PR_SENT_REPRESENTING_EMAIL_ADDRESS_UNICODE, cbdata->ownerid);
-	set_value (PR_SENDER_NAME_UNICODE, cbdata->username);
-	set_value (PR_SENDER_ADDRTYPE_UNICODE, cbdata->useridtype);
-	set_value (PR_SENDER_EMAIL_ADDRESS_UNICODE, cbdata->userid);
+	if (cbdata->ownername && cbdata->owneridtype && cbdata->ownerid) {
+		set_value (PR_SENT_REPRESENTING_NAME_UNICODE, cbdata->ownername);
+		set_value (PR_SENT_REPRESENTING_ADDRTYPE_UNICODE, cbdata->owneridtype);
+		set_value (PR_SENT_REPRESENTING_EMAIL_ADDRESS_UNICODE, cbdata->ownerid);
+	}
+
+	if (cbdata->username && cbdata->useridtype && cbdata->userid) {
+		set_value (PR_SENDER_NAME_UNICODE, cbdata->username);
+		set_value (PR_SENDER_ADDRTYPE_UNICODE, cbdata->useridtype);
+		set_value (PR_SENDER_EMAIL_ADDRESS_UNICODE, cbdata->userid);
+	}
 
 	flag32 = cbdata->msgflags;
 	set_value (PR_MESSAGE_FLAGS, &flag32);
@@ -1523,8 +1528,10 @@ exchange_mapi_cal_utils_write_props_cb (
 		flag32 = cbdata->appt_seq;
 		set_named_value (PidLidAppointmentSequence, &flag32);
 
-		set_named_value (PidLidCleanGlobalObjectId, cbdata->cleanglobalid);
-		set_named_value (PidLidGlobalObjectId, cbdata->globalid);
+		if (cbdata->cleanglobalid)
+			set_named_value (PidLidCleanGlobalObjectId, cbdata->cleanglobalid);
+		if (cbdata->globalid)
+			set_named_value (PidLidGlobalObjectId, cbdata->globalid);
 
 		flag32 = cbdata->resp;
 		set_named_value (PidLidResponseStatus, &flag32);
