diff -up evolution-mapi-0.28.3/src/libexchangemapi/exchange-mapi-connection.c.copymove-doublefree evolution-mapi-0.28.3/src/libexchangemapi/exchange-mapi-connection.c
--- evolution-mapi-0.28.3/src/libexchangemapi/exchange-mapi-connection.c.copymove-doublefree	2013-01-23 17:55:47.672078463 +0100
+++ evolution-mapi-0.28.3/src/libexchangemapi/exchange-mapi-connection.c	2013-01-23 17:56:06.272305064 +0100
@@ -2505,11 +2505,11 @@ mapi_move_items (mapi_id_t src_fid, mapi
 	result = TRUE;
 
 cleanup:
-	talloc_free (mem_ctx);
 	mapi_id_array_release(&msg_id_array);
 	mapi_object_release(&obj_folder_dst);
 	mapi_object_release(&obj_folder_src);
 	mapi_object_release(&obj_store);
+	talloc_free (mem_ctx);
 
 	return result;
 }
